### DISCUSSION TITLE
Unit Testing Azure Functions

### DISCUSSION THREAD

**Discussion by Ansiem** [Azure/azure-functions-sql-extension#268] [Q&A]
Hi there,

I am currently working on a project which contains multiple Azure functions and I am wondering if there is a way to Unit test the following Azure functions:

1) A Function with a Blob Storage trigger, this function gets triggered whenever a blob gets uploaded or updated. The output I'd like to test is if the function has added a row to a table in my MSSQL database.

2) A Function with a SQL Binding as a trigger, this function gets triggered whenever a row gets inserted in the previously mentioned table. I'd like to test if the function has added a message to a queue in my Storage Account.

3) A Function that gets triggered by a message in a specific queue, the output I'd like to test is if this function has placed a message in another queue (Storage Account).

The issue I am facing is that I don't know how to write unit test for these specifc Azure Functions with their triggers and outputs.

**Comment by Charles-Gagnon** [comment]
We have integration tests you can look at here : https://github.com/Azure/azure-functions-sql-extension/tree/main/test/Integration

These validate the sample functions - e.g. https://github.com/Azure/azure-functions-sql-extension/tree/main/samples/samples-csharp

1. We don't have a direct example of this, but we have a QueueTrigger test which is very similar https://github.com/Azure/azure-functions-sql-extension/blob/main/test/Integration/SqlOutputBindingIntegrationTests.cs#L134 (using https://github.com/Azure/azure-functions-sql-extension/blob/main/samples/samples-csharp/OutputBindingSamples/QueueTriggerProducts.cs)
2. We don't currently support Sql triggers. There's a PR out for adding that support but there's some architectural things we're working through before that will be ready to go in : https://github.com/Azure/azure-functions-sql-extension/pull/226
3. This doesn't sound like it involves SQL bindings at all, I'd suggest looking at tests for the Queue trigger/bindings for this
